### PR TITLE
First submission - may modify in the future.

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

**Comment by LukeWood** [comment]
Name: Luke Wood
SMU ID: 46802422

File: ex01.ex [hunk 1]
```diff
@@ -86,7 +45,9 @@ defmodule Ex01 do
   # function. The examples below will make this clearer :)               #
   ########################################################################
 
-  create_adder = your_anonymous_function(1)
+  create_adder = fn(diff) ->
+    fn(x) -> x + diff end
```
Comment (pragdave) [inline ex01.ex:49]: I like the naming of the parameter

File: ex02.ex [hunk 2]
```diff
@@ -3,62 +3,35 @@ Code.load_file "support.ex"
...
-  # 2.1:  5 points #
   ########################################################################
   # Write an anonymous function that takes two numbers and returns a two #
   # element list. The first element of the list will be the sum of the   #
   # numbers, and second should be the difference                         #
   ########################################################################
 
-  list2a = your_anonymous_function(1, 2)
-
+  list2a = fn (x, y) -> [x + y, x - y] end
```
Comment (pragdave) [inline ex02.ex:13]: I like the spacing you've used here.

If there were also multiplictive operators in the expressions, you could leave out the spacing around them, so I'd personally write:

~~~
a = 1 + b
~~~

and

~~~
a = 1 + 2*b
~~~

File: ex02.ex [hunk 3]
```diff
@@ -3,62 +3,35 @@ Code.load_file "support.ex"
...
 
-  ##################
-  # 2.2:  5 points #
   ##################################
   # Do the same using the & syntax #
   ##################################
 
-  list2b = your_anonymous_function(1, 2)
-
+  list2b = &([&1 + &2, &1 - &2])
```
Comment (pragdave) [inline ex02.ex:21]: No grade deduction, buit you can also leave off the parens here, so the idiomatic way would be to write

~~~ elixir
list2b = &[ &1 + &2, &1 - &2 ]
~~~

File: ex02.ex [hunk 4]
```diff
@@ -3,62 +3,35 @@ Code.load_file "support.ex"
...
   ################################################################
   # Write a function that uses pattern matching to return true   #
   # if the first two elements of a list are equal                #
   ################################################################
 
-  first2equal = your_anonymous_function([])
-
-
+  first2equal = fn
+    ([x, x | _t]) -> true
```
Comment (pragdave) [inline ex02.ex:31]: Although they don't change the meaning, I would avoid the parens here: they made me have to read it twice to see if I was missing something subtle.

The pattern is simply a list, so you can just write it as a list.

File: ex03.ex [hunk 5]
```diff
@@ -52,20 +13,15 @@ defmodule Ex03 do
...
 
-  def odd_even . . . "your code"
-
-
-  ##############################################################################
-  # 3.2:  5 points #
-  ##################
+  def odd_even([x | t]) when Integer.is_even(x), do: [:even | odd_even(t)]
+  def odd_even([x | t]) when Integer.is_odd(x), do: [:odd | odd_even(t)]
+  def odd_even(x), do: x
```
Comment (pragdave) [inline ex03.ex:20]: ### -2

I know you were trying to be thorough by adding this third clause, but by doing so you've actually broken the code.

If I called this function with `[ 1, :cat ]`, the result would be `[:odd, :cat ]`. I'd not get an error. But the description is explicit that the function should take a list of integers. 

One of the golden rules we talked about in the first week is that code should fail if given bad inputs, and not handle it.

Comment (LukeWood) [reply ex03.ex:20]: Good point - thanks

File: ex03.ex [hunk 6]
```diff
@@ -74,14 +30,11 @@ defmodule Ex03 do
...
   """
 
-  def list_contains . .. "your code"
-
-  ##############################################################################
-  # 3.3:  5 points #
-  ##################
+  def list_contains([x | _t], x), do: true
+  def list_contains([_1 | t], x), do: list_contains(t, x)
+  def list_contains(_, _x), do: false
```
Comment (pragdave) [inline ex03.ex:37]: Again this is too permissive, as it allow me to write

~~~ elixir
list_contains(:wombat, 1)
~~~

I'd suggest the last function head should be

~~~ elixir
def list_contains([], _x), do: false
~~~

I'm not docking any marks—the point was already made in 3.1

File: ex03.ex [hunk 7]
```diff
@@ -101,9 +54,9 @@ defmodule Ex03 do
 
   """
 
-  def list_equal . . . "your code"
-
-
+  def list_equal([], []), do: true
+  def list_equal([x | t1], [x | t2]), do: list_equal(t1, t2)
+  def list_equal(_, _), do: false
```
Comment (pragdave) [inline ex03.ex:59]: You know what I'm going to say... :)

File: ex03.ex [hunk 8]
```diff
@@ -126,31 +79,22 @@ defmodule Ex03 do
...
-
-      iex> Ex03.won { :o, :x, 3,   :x, :o, 6,   :x, :o, 9 }
-      false
-
-  Think a little about a nice way to lay this code out.
   """
 
-  def won . . . "your code"
-
+  def won({x, x, x, _4, _5, _6, _7, _8, _9}), do: x
```
Comment (pragdave) [inline ex03.ex:89]: I like the numbering of the unused parameters.

Here's a suggestion: you could lay this out with all the parameters aligned, which would make it even clearer:

~~~ elixir
def won({ x,  x,  x, _4, _5, _6, _7, _8, _9}), do:  x
def won({_1, _2, _3,  x,  x,  x, _7, _8, _9}), do:  x
def won({_1, _2, _3, _4, _5, _6,  x,  x,  x}), do:  x
def won({ x, _2, _3,  x, _5, _6,  x, _8, _9}), do:  x
def won({_1,  x, _3, _4,  x, _6, _7,  x, _9}), do:  x
def won({_1, _2,  x, _4, _5,  x, _7, _8,  x}), do:  x
def won({ x, _2, _3, _4,  x, _6, _7, _8,  x}), do:  x
def won({_1, _2,  x, _4,  x, _6,  x, _8, _9}), do:  x
def won(_), do: false
~~~

File: ex04.ex [hunk 9]
```diff
@@ -27,23 +23,14 @@ defmodule Ex04 do
...
       iex> Ex04.reverse [ 5, 4, 3, 2, 1 ]
       [ 1, 2, 3, 4, 5 ]
 
   """
-  def reverse . . . "your code"
-
-  ##############################################################################
-  # 4.2:  5 points #
-  ##################
+  def reverse(x), do: reduce(x, [], &([&1 | &2]))
```
Comment (pragdave) [inline ex04.ex:33]: Again, no need for parens.

File: ex04.ex [hunk 10]
```diff
@@ -75,10 +66,14 @@ defmodule Ex04 do
   return value will be the thing you have to manipulate.
   """
 
-  def even_odd . . . "your code"
-
-
-
+  def even_odd(x) do
```
Comment (pragdave) [inline ex04.ex:69]: ### -1

You clearly get the code, so I'm going to focus on the more detailed stuff here. In particular, we have a six line function, which is pushing it. It has a local variable, which is also a warning sign.

Never be afraid to split into functions:

~~~ elixir
  def even_odd(list) do
    reduce(list, { [], [] }, &partition_even_odd/2)
    |> reverse_two_lists
  end

  defp partition_even_odd(x, { es, os }) when Integer.is_even(x), do: {[x | es], os}
  defp partition_even_odd(x, { es, os }) when Integer.is_odd(x),  do: {es, [x | os]}

  defp reverse_two_lists({ evens, odds }) do
    { reverse(evens), reverse(odds) }
  end
~~~

**Comment by pragdave** [comment]
Really nice work. You clearly get this, so treat my comments more as a gentle nudge in the right direction rather than any kind of "this is wrong".

I hate to do this, but I'm going to have to deduct 2 marks for not submitting from a branch...

**Pull request closed without merging**